docs/cpp_build_systems.md

```diff
@@ -0,0 +1,350 @@
+# How Protobuf supports multiple C++ build systems
+
+This document explains how the Protobuf project supports multiple C++ build
+systems.
+
+## Background
+
+Protobuf uses [Bazel](https://bazel.build) as the source of truth for building
+the Protobuf C++ runtime and Protobuf compiler[^historical_sot]. However, there
+are several different build systems in common use for C++, each one of which
+requires essentially a complete copy of build definitions.
+
+[^historical_sot]:
+  On a historical note, prior to its [release as Open Source
+  Software](https://opensource.googleblog.com/2008/07/protocol-buffers-googles-data.html),
+  the Protobuf project was developed using Blaze, Google's internal build
+  system, and predecessor to Bazel (the vast majority of Google's contributions
+  continue to be developed this way). The Open Source Protobuf project, however,
+  historically used Autoconf as the source of truth for C++ build correctness.
+  Over time, other build systems (including Bazel) have been added, thanks in
+  large part to substantial contributions from the Open Source community. Since
+  the Protobuf project deals with multiple languages (all of which ultimately
+  rely upon C++, for the Protobuf compiler), Bazel is a natural choice for a
+  project-wide source of truth -- in fact, Bazel (and its predecessor, Blaze)
+  were designed in large part to support exactly this type of rich,
+  multi-language build.
+
+Currently, C++ Protobuf is buildable with Bazel, Autotools, and CMake. Each of
+these build systems has different semantics and structure, but share in common
+the list of files needed to build the runtime (and compiler).
+
+## Design
+
+### Extracting information from Bazel
+
+Bazel's Starlark API provides [aspects](https://bazel.build/rules/aspects) to
+traverse the build graph, inspect build rules, define additional actions, and
+expose information through
+[providers](https://bazel.build/rules/rules#providers). For example, the
+`cc_proto_library` rule uses an aspect to traverse the dependency graph of
+`proto_library` rules, and dynamically attaches actions to generate C++ code
+using the Protobuf compiler and compile using the C++ compiler.
+
+In order to support multiple build systems, the goal is to define the overall
```

Comment:
"the goal is to define the overall
build structure" --> "the overall build structure is defined"

Reply:
Done.